src/rust/engine/process_execution/src/docker.rs

```diff
@@ -450,28 +467,75 @@ impl<'a> CapturedWorkdir for CommandRunner<'a> {
     &'s self,
     _context: &'c Context,
     _workdir_path: &'w Path,
-    (container_id, sandbox_path_in_container): Self::WorkdirToken,
+    (container_id, working_dir): Self::WorkdirToken,
     req: Process,
     _exclusive_spawn: bool,
   ) -> Result<BoxStream<'r, Result<ChildOutput, String>>, String> {
     let docker = self.docker.get().await?;
 
-    let working_dir = req
-      .working_directory
-      .map(|relpath| Path::new(&sandbox_path_in_container).join(relpath))
-      .unwrap_or_else(|| Path::new(&sandbox_path_in_container).to_path_buf())
-      .into_os_string()
-      .into_string()
-      .map_err(|s| {
-        format!("Unable to convert working directory due to non UTF-8 characters: {s:?}")
-      })?;
-
     Command::new(req.argv)
       .env(req.env)
       .working_dir(working_dir)
       .spawn(docker, container_id)
       .await
   }
+
+  async fn prepare_workdir_for_capture(
+    &self,
+    _context: &Context,
+    _workdir_path: &Path,
+    (container_id, working_dir): Self::WorkdirToken,
+    req: &Process,
+  ) -> Result<(), String> {
+    // Docker on Linux will frequently produce root-owned output files in bind mounts, because we
+    // do not assume anything about the users that exist in the image.
+    if matches!(
+      Platform::current()?,
+      Platform::Macos_x86_64 | Platform::Macos_arm64
+    ) {
+      return Ok(());
+    }
+
+    let docker = self.docker.get().await?;
+
+    let args = ["chmod", "a+r", "-R"]
```

Comment:
There's no recursive chmod function in Rust I guess?

Reply:
This will give read permissions, but I still won't be able to delete these files if I try to clean up the host side of the bind mount.

Reply:
Right 🤦

Reply:
If you did already see this as an issue I'm mildly concerned. That implies there is a force pushing us to ship broken things and not slow down. Its good this got flushed out in review and bugged, but we already have general concerns over `~/.cache/pants` size voiced by users and now we are making new issues for `/tmp`.

Reply:
I completely agree. I think the evil here is we're desperately trying to ship a blog post and not quality code.

Reply:
Would this counsel renaming `docker_environment` to `experimental_docker_environment` (besides listing the caveats)?

Reply:
> Would this counsel renaming `docker_environment` to `experimental_docker_environment` (besides listing the caveats)?

Possibly, but we already call the entire environments feature experimental, since to give any environment a name, users need to set `[environments-preview].names` for their environments. Some of the [other environment issues](https://github.com/pantsbuild/pants/issues/17355) are arguably as important as those specific to docker (not applying environments to `fix/fmt/lint/check` in particular).